misk/web/@misk/components/src/components/TopbarComponentDimensionAware.tsx

```diff
@@ -0,0 +1,388 @@
+  margin: 5px 0 10px 0;
+`
+
+const MiskBanner = styled.span`
+  background-color: ${props => props.color} !important;
+  color: ${color.accent} !important;
+  text-align: center;
+  font-weight: 600;
+  padding: 5px 10px;
+  position: fixed !important;
+  width: 100%;
+  top: 70px;
+  left: 0px;
+  z-index: 1010 !important;
+
+  a {
+    font-weight: 300;
+    color: ${color.accent};
+    text-decoration: underline;
+    letter-spacing: 1px;
+    &:hover {
+      color: ${color.white};
+      text-decoration: underline;
+    }
+  }
+`
+
+const environmentToColor = (environment: Environment) => {
+  switch (environment) {
+    case Environment.DEVELOPMENT:
+      return color.blue
+    case Environment.TESTING:
+      return color.indigo
+    case Environment.STAGING:
+      return color.green
+    case Environment.PRODUCTION:
+      return color.red
+  }
+}
+
+export class TopbarComponentDimensionAware extends React.Component<
+  IDimensionAwareProps & ITopbarProps,
+  {}
+> {
+  public state = {
+    isOpen: false
+  }
+
+  public render() {
+    const { isOpen } = this.state
+    const {
+      environment,
+      environmentBannerVisible,
+      error,
+      homeName,
+      homeUrl,
+      links,
+      width,
+      status
+    } = this.props
+    return (
+      <MiskNavbar>
+        <MiskNavbarButton onClick={this.handleClick}>
+          <MiskNavbarIcon
+            iconSize={32}
+            icon={isOpen ? IconNames.CROSS : IconNames.MENU}
+          />
+        </MiskNavbarButton>
+        <ResponsiveContainer>
+          <MiskNavbarGroup align={Alignment.LEFT} className="bp3-dark">
+            {this.renderNavbarHome(homeName, homeUrl)}
+            {this.NavbarItems(environment)
+              .slice(0, Math.floor(Math.min(width - 300, 1800) / 400))
+              .map(item => item)}
+          </MiskNavbarGroup>
+        </ResponsiveContainer>
+        <MiskCollapse isOpen={isOpen} keepChildrenMounted={true}>
+          <MiskMenu>
+            <ResponsiveContainer>
+              <MiskMenuNavbarItems>
+                <FlexContainer>
+                  {this.NavbarItems(environment).map(item => (
+                    <span key={item.key} onClick={this.handleClick}>
+                      {item}
+                    </span>
+                  ))}
+                </FlexContainer>
+              </MiskMenuNavbarItems>
+              {links ? (
+                this.renderMenuCategories(links)
+              ) : (
+                <ErrorCalloutComponent error={error} />
+              )}
+            </ResponsiveContainer>
+          </MiskMenu>
+        </MiskCollapse>
+        {this.renderBanner(status, environment, environmentBannerVisible)}
+      </MiskNavbar>
+    )
+  }
+
+  private NavbarItems(environment?: Environment) {
+    console.log(this.props.navbarItems)
+    return [this.renderNavbarEnvironment(environment)].concat(
+      this.props.navbarItems
+        ? this.props.navbarItems.map(item => (
```

Comment:
Why are navbar items so heterogeneous to need a component to marshall them into spans?

Reply:
Long term we want the ability to arbitrarily inject potentially complex full elements such as search bar and others that may or may not ship in the same `@misk/components` package.